include/deal.II/base/patterns.h

```diff
@@ -661,6 +661,98 @@ namespace Patterns
   };
 
 
+
+  /**
+   * This pattern matches comma-separated values of arbitrary types. Each type
+   * has to match a pattern given to the constructor.
+   *
+   * The constructor expects a vector of Patterns, and optionally an
+   * std::string, specifying the separator to use when parsing the Tuple from a
+   * string.
+   */
+  class Tuple : public PatternBase
+  {
+  public:
+    /**
+     * Constructor. Specify each pattern that the Tuple should contain.
+     *
+     * Optionally specify a separator string.
+     */
+
+    /**
+     * Constructor.
+     * @param patterns The pattern each object of the Tuple should match
+     * @param separator An optional string used to delimit each element
+     */
+    Tuple (const std::vector<std::unique_ptr<PatternBase> > &patterns,
```

Comment:
Have you thought about using a `std::tuple<...>` as argument here, where each element of the tuple needs to be a `Pattern`?

Reply:
Well, since every element needs to be a Pattern anyway, why not use a uniform data container?

Reply:
If I used a Tuple, I'd have to check that every element is a Pattern...

Reply:
We typically clone these patterns anyway. In any case, I could create the pattern argument then as follows:
```
  prm.declare_entry ("foo", "...",
                                 Patterns::Tuple(std::make_tuple(Patterns::Integer(3,5),
                                                                                     Patterns::Double())));
```
Though that suggests that maybe a variadic template may be even more appealing, because then I could write
```
  prm.declare_entry ("foo", "...",
                                 Patterns::Tuple(Patterns::Integer(3,5),
                                                           Patterns::Double()));
```
You would get this by writing the constructor as
```
  template <typename... Patterns>
  Tuple::Tuple (Patterns... patterns)
    :
    vector_of_unique_ptr_member_variable_whose_name_I_forgot
                 ({(patterns.clone())...})
    {}
```

Reply:
How would you specify the optional separator argument?

Reply:
Since the variadic template argument apparently needs to be last, you need to make the separator first. This is awkward because then you can't give it a default value. But you can do this:
```
  template <typename... Patterns>
  Tuple::Tuple (char separator,
                        Patterns... patterns)
    :
    vector_of_unique_ptr_member_variable_whose_name_I_forgot
                 ({(patterns.clone())...}),
    separator (separator)
    {}

  template <typename... Patterns>
  Tuple::Tuple (Patterns... patterns)        // sep=','
    :
    // just forward to the other constructor
    Tuple(',', patterns...)
    {}
```

Reply:
ah. neat. I'll try that. thanks!

Reply:
You might then also put a `static_assert` in place that checks that all the `Patterns...` are in fact derived from `PatternBase`. Something similar to
```
static_assert(internal::TemplateConstraints::all_true<std::is_base_of<PatternBase,Patterns>::value...>::value, "Not all the arguments are derived from PatternBase!");
```
should work.